src/contrib/mmap_cache.rst

```diff
@@ -0,0 +1,506 @@
+with a defined size. The current slot size is 40 bytes. Since the slot size is
+not given in the header the ``major_vno`` or ``minor_vno`` must be changed if
+the slot size is changed. Each data record for an individual user or group (see
+below) can occupy one or more consecutive slots.
+
+At startup the data table is initialized with ``0xff`` (all bits set) which is
+treated as invalid value by the memory cache code independent if 1, 4 or 8 bytes
+are read.
+
+The Free Table
+--------------
+
+The memory area is treated as bit-field. Each bit corresponds to a slot in the
+data table starting with the highest bit (``0x80``) of the first byte in the
+free table.
+
+At startup the free table is initialized with ``0x00`` (no bit set) which
+corresponds to a completely empty data table.
+
+The Hash Table
+--------------
+
+The hash table it treated as an array of ``uint32_t`` values. The index used to
+access the array is the hash value calculated from the user or group name or ID
+and the hash value is the slot number in the data table where the data record
+for the object starts.
+
+At startup the hash table is initialized with ``0xff`` which indicates that
+there is no data available for the given hash.
+
+Sizes of the tables
+-------------------
+
+From the previous sections it became clear that although the sizes of the data,
+free and hash tables are handled independently they are connected. Of course the
+free table should have at least so many bit as there are slots in the data
+table. But having more would be a waste of space (although not as much as having
+lesser bits). So number of bits and slots should match.
+
+The size of the data table should be large enough to help the amount of expected
+data where expected do not necessarily mean e.g. all possible users but the
+amount of data needed on the host running SSSD to work efficiently. It should be
+large enough so that data in active use does not have to be overwritten
+constantly with other data. So let's say there are *N* data elements in
+the cache. Depending on the slot size and the type of data one slot might not be
+sufficient to hold a data element of average size so there is a payload which
+says how many slots are used typically for a data element. See below for a
+discussion about the payload of the current data types. For the size of the data
+table we now have *N* times the payload size.
+
+With the number of data elements *N* we can now define the size of the hash
+table as well. Since the data typically can be accessed either by name or ID we
+need 2 times *N* entries in the hash table array for the ideal case where no
+hash collisions occur. Since we always have to expect and handle hash collisions
+it won't help to use more memory for the hash table. On the other hand hash
+collisions should be avoided where possible because they make the handling of
+the memory cache more complex and slower. Since using less than 2 times *N*
```

Comment:
s/because they make the handling of the memory cache more complex and slower/because handling of the memory caches becomes slower and more complex/